### DISCUSSION TITLE
Specifying extensions

### DISCUSSION THREAD

**Discussion by globalflea** [serverlessworkflow/specification#542] [General]
Is there a way to specify extensions beyond the workflow specifications? I.e. is there a way to specify other stuff i may want to add but ignored by workflow engines that do not recognise my extended features?

 E.g if i want to include DSL rules for the rule engine but tied to a workflow or any state in a workflow; or even embed code snippets??

**Comment by tsurdilo** [comment]
You can use metadata - https://github.com/serverlessworkflow/specification/blob/main/specification.md#workflow-metadata
Let me know if that works

**Reply by globalflea** [reply]
i was thinking more of a set of tags to mark it as an extension and another that classify the extension? in this way, i can have multiple extensions (of different types).  Does this make sense?

**Reply by tsurdilo** [reply]
ok just so i understand, we just added a fix for current extensions, so we can actually reference them inside workflow, for example:

    {
      "extensions": [
        {
          "extensionId": "workflow-ratelimiting-extension",
          "path": "file://myextensions/ratelimiting.yml"
        },
        {
          "extensionId": "workflow-kpi-extension",
          "path": "file://myextensions/kpi.yml"
        }
      ]
    }

(see bottom of section https://github.com/serverlessworkflow/specification/blob/main/specification.md#workflow-definition-structure)
the way i currently understand it you would like to have for example:

      {
      "extensions": [
        {
          "name": "myratelimitextension",
          "extensionId": "workflow-ratelimiting-extension",
          "path": "file://myextensions/ratelimiting.yml"
        },
        {
          "name": "mykpiextension",
          "extensionId": "workflow-kpi-extension",
          "path": "file://myextensions/kpi.yml"
        }
      ]
     }

and then in a workflow state for example:

    {
       "name": "myState",
       "type": "operation",
       ....
       "extensionRefs": ["myratelimitextension", "mykpiextension"]
    }

something like that ?
If I'm off, please show sample of what you think it should look like.

**Reply by globalflea** [reply]
My bad, didn't know you have included the extension mechanism already.  Thanks for pointing it out.  Think your example above would work nicely.  

May be good to include some sort of a namespace or category - to indicate if it's workflow extension, or other classes of extension - eg. rule (or rule-engine), constraints (e.g resource requirements), optimisation ... something the user can specify

**Reply by tsurdilo** [reply]
Ok. Would be nice to see an example of what you are trying to do in order to figure out how best to add a feature.

**Reply by globalflea** [reply]
how about:
  {
  "extensions": [
    {
      "name": "myratelimitextension",
      **"namespace": "org.cncf.serverless-workflow"**
      "extensionId": "workflow-ratelimiting-extension",
      "path": "file://myextensions/ratelimiting.yml"
    },

 where "namespace" (or something better) is used to group the extensions into specific groups or vendor specific extensions?

Also, by specifying the "extensionId" and "path", you are forcing the vendors to put their extensions in a separate files.  Perhaps it may be better to leave the rest of the parameters up to the individual vendors?  so maybe these two parameters can be optional?  Either these two are specified or a Json object (containing vendor specs) may be used?

**Reply by tsurdilo** [reply]
i like it! could you please open issue and ill work on it
